### PR TITLE
Fix double free remapper->typmodmap in TeardownUDPIFCInterconnect()

### PR DESCRIPTION
**Pull request by interma**
Usually, the `typmodmap` field is NULL. And it has a valid value if need the "record type cache", example:
```
create table t as select i as id from generate_series(1,8) i; 
create type myudt as (i int, c char);
-- the following query will generate "record type cache"
select row((1,'a')::myudt), row((2,'b')::myudt, 2) from t;
```

When the similar query raises a FATAL, it runs into the following routine:
```
FATAL->AbortTransaction()
    AtAbort_Portals()
        MemoryContextDeleteChildren()     // delete ExecutorState memory context
    ResourceOwnerRelease()
        cleanup_interconnect_handle()
            TeardownUDPIFCInterconnect()
                DestroyTupleRemapper()          // pfree remapper->typmodmap
```
But it may cause SIGSEGV, reason: 
`remapper->typmodmap` is alloced in `MotionLayerMemCtxt`, but the memory context has been deleted in `AtAbort_Portals()` (it's the child of `ExecutorState`).

This PR fixes it: 
Align the used memory context: `remapper` and its field `typmodmap` and `field_remapinfo` are alloced in the same content `remapper->mycontext` (it's usually `InterconnectContext`).

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
Thanks a lot for this hard memory bug fixing. The commit message is well written. Would you please also consider adding comments on test and code？

**Comment by interma** [comment]
> Thanks a lot for this hard memory bug fixing. The commit message is well written. Would you please also consider adding comments on test and code？

Sure, thanks Zhenghua.
I will add them.

File: src/backend/cdb/motion/tupleremap.c [hunk 1]
```diff
@@ -224,10 +225,12 @@ TRHandleTypeLists(TupleRemapper *remapper, List *typelist)
 	ListCell   *cell;
 	int			mapsize = remapper->typmodmapsize + list_length(typelist);
 
+	MemoryContext oldcontext = MemoryContextSwitchTo(remapper->mycontext);
```
Comment (kainwen) [inline src/backend/cdb/motion/tupleremap.c:228]: Please add a comment on what `remapper->mycontext` will be and its parent context.

**Review by kainwen — APPROVED** [review]
LGTM. Good job!

File: src/backend/cdb/motion/tupleremap.c (hunk 1 above)
Comment (interma) [reply src/backend/cdb/motion/tupleremap.c:229]: fixed

**Pull request merged**